src/phoenix/experimental/evals/functions/executor.py

```diff
@@ -0,0 +1,338 @@
+        join_task = asyncio.create_task(queue.join())
+        termination_signal_task = asyncio.create_task(termination_signal.wait())
+        done, pending = await asyncio.wait(
+            [join_task, termination_signal_task], return_when=asyncio.FIRST_COMPLETED
+        )
+        if termination_signal_task in done:
+            # Cancel all tasks
+            if not join_task.done():
+                join_task.cancel()
+            if not producer.done():
+                producer.cancel()
+            for task in consumers:
+                if not task.done():
+                    task.cancel()
+
+        if not termination_signal_task.done():
+            termination_signal_task.cancel()
+
+        # reset the SIGTERM handler
+        signal.signal(signal.SIGINT, signal.SIG_DFL)  # reset the SIGTERM handler
+        return outputs
+
+    def run(self, inputs: Sequence[Any]) -> List[Any]:
+        return asyncio.run(self.execute(inputs))
+
+
+class SyncExecutor(Executor):
+    """
+    Synchronous executor for generating outputs from inputs using a given generation function.
+
+    Args:
+        generation_fn (Callable[[Any], Any]): The generation function that takes an input and
+        returns an output.
+
+        tqdm_bar_format (Optional[str], optional): The format string for the progress bar. Defaults
+        to None.
+
+        max_retries (int, optional): The maximum number of times to retry on exceptions. Defaults to
+        10.
+
+        exit_on_error (bool, optional): Whether to exit execution on the first encountered error.
+        Defaults to True.
+
+        fallback_return_value (Union[Unset, Any], optional): The fallback return value for tasks
+        that encounter errors. Defaults to _unset.
+    """
+
+    def __init__(
+        self,
+        generation_fn: Callable[[Any], Any],
+        tqdm_bar_format: Optional[str] = None,
+        max_retries: int = 10,
+        exit_on_error: bool = True,
+        fallback_return_value: Union[Unset, Any] = _unset,
+    ):
+        self.generate = generation_fn
+        self.fallback_return_value = fallback_return_value
+        self.tqdm_bar_format = tqdm_bar_format
+        self.max_retries = max_retries
+        self.exit_on_error = exit_on_error
+
+        self._TERMINATE = False
+
+    def _signal_handler(self, signum: int, frame: Any) -> None:
+        tqdm.write("Process was interrupted. The return value will be incomplete...")
+        self._TERMINATE = True
+
+    def run(self, inputs: Sequence[Any]) -> List[Any]:
+        signal.signal(signal.SIGINT, self._signal_handler)
+        outputs = [self.fallback_return_value] * len(inputs)
+        progress_bar = tqdm(total=len(inputs), bar_format=self.tqdm_bar_format)
+
+        for index, input in enumerate(inputs):
+            try:
+                for attempt in range(self.max_retries + 1):
+                    if self._TERMINATE:
+                        return outputs
+                    try:
```

Comment:
The `SyncExecutor` class has been updated to mirror the `AsyncExecutor` in it's error handling behavior, retrying tasks that raise exceptions that bubble up to the `Executor` up to `max_retries` times. We ultimately intend for this mechanism to replace `tenacity` retries